lib/src/cli/deployapps/baas_client.dart

```diff
@@ -0,0 +1,294 @@
+    // process the reset token, tokenId, username and password
+    if (password.includes("realm_tests_do_reset")) {
+      return { status: 'success' };
+    }
+    // will not reset the password
+    return { status: 'fail' };
+  };''';
+
+  final String _baseUrl;
+  final String? _clusterName;
+  final String _appSuffix;
+  final Map<String, String> _headers;
+
+  late String _groupId;
+
+  BaasClient._(String baseUrl, [this._clusterName])
+      : _baseUrl = "$baseUrl/api/admin/v3.0",
+        _headers = <String, String>{'Accept': 'application/json'},
+        _appSuffix = "-$_clusterName";
+
+  static Future<BaasClient> docker(String baseUrl) async {
+    final BaasClient result = BaasClient._(baseUrl);
+
+    await result._authenticate("local-userpass", '{ "username": "unique_user@domain.com", "password": "password" }');
+
+    dynamic groupDoc = await result._get("auth/profile");
+    result._groupId = (groupDoc['roles'] as List<dynamic>)[0]['group_id'] as String;
+
+    print("Current GroupID ${result._groupId}");
+
+    return result;
+  }
+
+  static Future<BaasClient> atlas(String baseUrl, String cluster, String apiKey, String privateApiKey, String groupId) async {
+    final BaasClient result = BaasClient._(baseUrl, cluster);
+
+    await result._authenticate("mongodb-cloud", '{ "username": "$apiKey", "apiKey": "$privateApiKey" }');
+
+    result._groupId = groupId;
+
+    return result;
+  }
+
+  Future<Map<String, BaasApp>> getOrCreateApps() async {
+    final result = <String, BaasApp>{};
+    var apps = await _getApps();
+    if (apps.isNotEmpty) {
+      for (var app in apps) {
+        result[app.name] = app;
+      }
+    } else {
+      final defaultApp = await _createApp("flexible");
+
+      result[defaultApp.name] = defaultApp;
+    }
+
+    return result;
+  }
+
+  Future<List<BaasApp>> _getApps() async {
+    final apps = await _get("groups/$_groupId/apps") as List<dynamic>;
+    return apps
+        .map((dynamic doc) {
+          final name = doc['name'] as String;
+          if (!name.endsWith(_appSuffix)) {
+            return null;
+          }
+
+          final appName = name.substring(0, name.length - _appSuffix.length);
+          return BaasApp(doc['_id'] as String, doc['client_app_id'] as String, appName);
+        })
+        .where((doc) => doc != null)
+        .map((doc) => doc!)
+        .toList();
+  }
+
+  Future<BaasApp> _createApp(String name) async {
+    print("Creating app $name");
+
+    final dynamic doc = await _post("groups/$_groupId/apps", '{ "name": "$name$_appSuffix" }');
+    final appId = doc['_id'] as String;
+    final clientAppId = doc['client_app_id'] as String;
+
+    final app = BaasApp(appId, clientAppId, name);
+
+    final confirmFuncId = await _createFunction(app, "confirmFunc", _confirmFuncSource);
+    final resetFuncId = await _createFunction(app, "resetFunc", _resetFuncSource);
+
+    enableProvider(app, "anon-user");
+    enableProvider(app, "local-userpass", '''{
```

Comment:
Have you considered doing a proper type, and use `json_serializable` to source generate the serialization code?

Reply:
Yeah - I thought about it, but felt it's an overkill considering we're mostly serializing hardcoded values. Almost none of the payloads or responses are reused meaning we get very little value from having them be strongly typed.